DECISION-RECORD.md

```diff
@@ -0,0 +1,292 @@
+
+- URL parameters
+- For exact matches: /items?state=active
+- For non-exact matches:
+  - Filters are composed by:
+    - The property or field name
+    - The operator such as eq, lte, gte
+    - The filter value
+
+Sources: 
+
+- https://www.moesif.com/blog/technical/api-design/REST-API-Design-Filtering-Sorting-and-Pagination/
+- https://docs.microsoft.com/en-us/graph/query-parameters#filter-parameter
+
+### LHS Brackets
+
+Retrieves all the items where the price is greater than or equal to 10, but less than or equal to 100
+
+`GET /items?price[gte]=10&price[lte]=100`
+
+#### Pros
+
+- Ease of use for clients. Many query string parsing libraries available to encode nested JSON objects into square brackets
+- Simple to parse on server side. The URL parameter key contains both the field name and operator. Easy to GROUP BY (property name, operator) without looking at the URL parameter values.
+- No need to escape special characters in the filter value when operator is taken as a literal filter term. 
+
+#### Cons
+
+- More work on server side to parse and group the filters. Custom URL parameter binder or parser to split the query string key into two components (field name, operator) may be needed
+- Special characters in variable names can be awkward. Custom binder to split the query string key into two components may be needed
+- Hard to manage custom combinational filters since there is an implicit AND among filters. How to manage OR filters?
+
+### RHS Colon
+
+Retrieves all the items where the price is greater than or equal to 10, but less than or equal to 100
+
+`GET /items??price=gte:10&price=lte:100`
+
+#### Pros
+
+- Easiest to parse on server side especially if duplicate filters are not supported. 
+  - No custom binders needed. 
+  - Many API frameworks already handle URL parameter arrays. 
+  - Multiple price filters will be under the same variable ‘price’ which may be a Sequence or Map. 
+
+#### Cons
+
+Literal values need special handling.
+
+Retrieve all items where the user_id is greater than 100. 
+
+`GET /items?user_id=gt:100 `
+
+However, if `gt:100` was a valid id, we could not find the items where the `user_id` equals `gt:100`
+
+### Search query param
+
+Retrieves all the items containing the terms red chair and the price is greater than or equal to 10 and less than or equal to 100
+
+`GET /items?q=title:red chair AND price:[10 TO 100]`
+
+- Adding filters and ranges with the search parameter
+  - Lucene syntax or ElasticSearch Simple Query Strings would be easy to support directly - if needed.
+
+#### Pros
+
+- Most flexible queries for API users
+- Almost no parsing required on backend, can pass directly to search engine or database
+
+#### Cons
+
+- Harder for beginners (Lucene syntax)
+- Full-text search doesn’t make sense for all resources. E.g. Fuzziness and term boosting doesn’t make sense for time series metric data.
+- Requires URL percent encoding i.e. use of cURL or Postman becomes more complicated.
+
+---
```

Comment:
This is fantastic research, but I think we can just settle for a shorter summary here and extract this into a different document

Reply:
This is one of the reasons I thought about multiple markdown docs.
I'd like to keep the detail here, at least initially, as I believe it would be useful to others.